main.go

```diff
@@ -140,55 +150,132 @@ func (d *MDNSDiscovery) StartSync(eventCB discovery.EventCallback, errorCB disco
 	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
 		defer close(queriesChan)
+		queryLoop(ctx, queriesChan)
+	}()
+	go func() {
+		for entry := range queriesChan {
+			if d.entriesChan != nil {
+				d.entriesChan <- entry
+			}
+		}
+	}()
+	d.cancelFunc = cancel
+	return nil
+}
 
-		disableIPv6 := false
-		// Check if the current network supports IPv6
-		mconn6, err := net.ListenMulticastUDP("udp6", nil, ipv6Addr)
+func queryLoop(ctx context.Context, queriesChan chan<- *mdns.ServiceEntry) {
+	for {
+		var interfaces []net.Interface
+		var conn connectivity
+		var wg sync.WaitGroup
+
+		interfaces, err := availableInterfaces()
 		if err != nil {
-			disableIPv6 = true
-		} else {
-			mconn6.Close()
+			goto NEXT
 		}
 
-		// We must check if we're connected to a local network, if we don't
-		// the subsequent mDNS query would fail and return an error.
-		mconn4, err := net.ListenMulticastUDP("udp4", nil, ipv4Addr)
-		if err != nil {
+		conn = checkConnectivity()
+		if conn.available() {
+			goto NEXT
+		}
+
+		wg.Add(len(interfaces))
+
+		for n := range interfaces {
+			go func(params *mdns.QueryParam) {
+				defer wg.Done()
+				mdns.Query(params)
+			}(makeQueryParams(&interfaces[n], conn, queriesChan))
+		}
```

Comment:
Urgh, this makes my brain twist :-)

what about:

```suggestion
		for n := range interfaces {
			params := makeQueryParams(&interfaces[n], conn, queriesChan)
			go func() {
				defer wg.Done()
				mdns.Query(params)
			}()
		}
```

so we instantiate the `params` in the loop and we do not need to pass it as `func(...)` parameter to create the closure.

Reply:
I often get caught in the go's loop variable lifetime mess, wasn't sure it is yet another case of that. Same for the below `netif :=` btw.
(i.e. when passing `range` value and getting same thing in all invocations)

Reply:
> (i.e. when passing range value and getting same thing in all invocations)

Yeah, this happens when you spawn goroutines inside the for loop:

```go
	for n := range elements {
		go func() {
			fmt.Println(n)  // <----
		}()
	}
```

It's a very unfortunate trap of go: the `n` pointed by the arrow is not a copy, but it's the same `n` in the for loop. With very high probability the goroutines will be executed after the for loop finishes, so they all get the last value of `n` (i.e. the len of `elements`). There is no guarantee on the execution order of the goroutines
https://go.dev/play/p/D9dMU-sDTAz

To avoid this there a two ways:
1. pass the for index as parameter of the function:
```go
	for n := range elements {
		go func(idx int) {
			fmt.Println(idx) // <--- this is a copy of n
		}(n)
	}
```
so it gets actually copied to the local variable `idx` or

2. make a new copy of the index variable and use that inside the goroutine
```go
	for n := range elements {
		idx := n
		go func() {
			fmt.Println(idx)
		}()
	}
```
in this case the `idx` is created new at each loop iteration (outside of the goroutine) so each goroutine reference his own copy of `idx`. I prefer this solution because I found it much more readable.